pkg/rotator/rotator_test.go

```diff
@@ -153,3 +162,207 @@ func TestEmptyIsInvalid(t *testing.T) {
+		createSecret(ctx, g, c, key)
+	}
+
+	stop, wg := StartTestManager(mgr, g)
+
+	// The reader (cache) will be initialized in AddRotator.
+	g.Expect(rotator.reader).ToNot(gomega.BeNil())
+
+	// Wait for it to populate
+	if !rotator.reader.WaitForCacheSync(stop) {
+		t.Fatal("waiting for cache to populate")
+	}
+
+	lst := &corev1.SecretList{}
+	err = rotator.reader.List(ctx, lst)
+	g.Expect(err).ToNot(gomega.HaveOccurred(), "listing secrets")
+
+	g.Expect(lst.Items).To(gomega.HaveLen(1), "expected only single secret")
+	g.Expect(lst.Items[0].Namespace).To(gomega.Equal(key.Namespace), "verifying secret namespace")
+	g.Expect(lst.Items[0].Name).To(gomega.Equal(key.Name), "verifying secret name")
+
+	close(stop)
+	wg.Wait()
+}
+
+func ensureCertWasGenerated(ctx context.Context, g *gomega.WithT, c client.Reader, secret *corev1.Secret) {
+	const timeout = 15 * time.Second
+	const interval = 50 * time.Millisecond
+	key := types.NamespacedName{
+		Namespace: secret.Namespace,
+		Name:      secret.Name,
+	}
+	g.Eventually(func() bool {
+		if err := c.Get(ctx, key, secret); err != nil {
+			return false
+		}
+
+		return len(secret.Data) > 0
+	}, timeout, interval).Should(gomega.BeTrue(), "waiting for certificate generation")
+}
+
+func ensureWebhookPopulated(ctx context.Context, g *gomega.WithT, c client.Reader, wh *admissionv1.ValidatingWebhookConfiguration) {
+	const timeout = 15 * time.Second
+	const interval = 50 * time.Millisecond
+	key := types.NamespacedName{
+		Name: wh.Name,
+	}
+	g.Eventually(func() bool {
+		if err := c.Get(ctx, key, wh); err != nil {
+			return false
+		}
+
+		if len(wh.Webhooks) == 0 {
+			return false
+		}
+
+		for _, vw := range wh.Webhooks {
+			if len(vw.ClientConfig.CABundle) == 0 {
+				return false
+			}
+		}
+		return true
+	}, timeout, interval).Should(gomega.BeTrue(), "waiting for webhook reconciliation")
+}
+
+func resetWebhook(ctx context.Context, g *gomega.WithT, c client.Client, wh *admissionv1.ValidatingWebhookConfiguration) {
+	key := types.NamespacedName{
+		Name: wh.Name,
+	}
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		if err := c.Get(ctx, key, wh); err != nil {
+			return err
+		}
+		for i := range wh.Webhooks {
+			wh.Webhooks[i].ClientConfig.CABundle = nil
+		}
+		return c.Update(ctx, wh)
+	})
+	g.Expect(err).NotTo(gomega.HaveOccurred(), "resetting webhook")
+}
+
+// createSecret creates an empty secret.
+// If the referenced namespace does not exist, it will be created.
+func createSecret(ctx context.Context, g *gomega.WithT, c client.Writer, secretKey types.NamespacedName) {
+	err := c.Create(ctx, &corev1.Namespace{
+		ObjectMeta: metav1.ObjectMeta{Name: secretKey.Namespace},
+	})
+	g.Expect(err).NotTo(gomega.HaveOccurred(), "creating namespace", secretKey.Namespace)
+
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: secretKey.Namespace,
+			Name:      secretKey.Name,
+		},
+	}
+	err = c.Create(ctx, secret)
```

Comment:
Does this need retry logic?

Reply:
In the context of the existing tests, I've taken care to provide a unique key for each call, so there shouldn't be any conflicts. This could change in the future, or if you run the tests multiple times with the same envtest instance (go test -count=2 ./pkg/rotator). As long as it's deterministic, do we need more?
I'm worried that if I ignore OnConflict errors, I wouldn't know the contents of the secret anymore, unless this helper morphs into a CreateOrUpdate helper.